### PR TITLE
3.next - Add Help shell

### PR DESCRIPTION
**Pull request by markstory**
This shell is a replacement for CommandListShell. It uses the provided CommandCollection to list the commands instead of scanning the application. This will allow it to play nicer with standalone applications.

I've not yet fixed the hardcoded `cake` in the help output. But I'll be doing that in my next pull request for the new console libraries.

Refs #10716

### CODE REVIEW

File: src/Console/CommandRunner.php [hunk 1]
```diff
@@ -157,11 +161,15 @@ protected function getShell(ConsoleIo $io, CommandCollection $commands, $name)
...
         }
-        $classOrInstance = $commands->get($name);
-        if (is_string($classOrInstance)) {
-            return new $classOrInstance($io);
+        $instance = $commands->get($name);
+        if (is_string($instance)) {
+            $instance = new $instance($io);
+        }
+        // TODO Should this be an interface?
+        if (method_exists($instance, 'setCommandCollection')) {
```
Comment (markstory) [inline src/Console/CommandRunner.php]: I'm contemplating doing this at the same time as propagating the root command name. Right now we have a few places we hard code `cake` when it should really reflect the root command name given to the `CommandRunner`.

Comment (lorenzo) [inline src/Console/CommandRunner.php]: keeping the root command name everywhere would be nice :)

Comment (dereuromark) [inline src/Console/CommandRunner.php]: Interface would sure be a cleaner way :+1:

Comment (markstory) [inline src/Console/CommandRunner.php]: Ok. I'll do both of those in a subsequent pull request :+1:

**Comment by lorenzo** [comment]
Good job @markstory

File: src/Shell/CommandListShell.php [hunk 2]
```diff
@@ -145,7 +146,7 @@ public function getOptionParser()
             'help' => 'Get the listing as XML.',
             'boolean' => true
         ])->addOption('version', [
-            'help' => 'Prints the currently installed version of CakePHP.',
+            'help' => 'Prints the currently installed version of CakePHP. (deprecated - use `cake --version` instead)',
```
Comment (ravage84) [inline src/Shell/CommandListShell.php]: I do not understand. If you want to see the help, you have to use `cake --version` instead? That does not make much sense to me.

Comment (markstory) [inline src/Shell/CommandListShell.php]: I was trying to document that `cake command_list --version` is deprecated.

**Pull request merged**